lib/queue.js

```diff
@@ -1,256 +1,527 @@
+ * @param {Number=} page.end End of query range for waiting/active/delayed queue types
+ * @param {Number=} page.size Number jobs to return for failed/succeeded (SET) types
+ * @param {Function=} callback Called with the equivalent of the returned promise.
+ * @return {Promise<Job[]>} Resolves to the jobs the function found.
+ */
+Queue.prototype.getJobs = function (type, page, cb) {
+  const promise = this._commandable().then((client) => {
+    const idsPromise = helpers.deferred();
+    switch (type) {
+      case 'failed':
+      case 'succeeded':
+        this._scanForJobs(this.toKey(type), '0', page.size, new Set(), idsPromise.defer());
+        break;
+      case 'waiting':
+      case 'active':
+        client.lrange(this.toKey(type), page.start, page.end, idsPromise.defer());
+        break;
+      case 'delayed':
+        client.zrange(this.toKey(type), page.start, page.end, idsPromise.defer());
+        break;
+      default:
+        throw new Error('Improper queue type');
+    }
+
+    return idsPromise;
+  }).then((ids) => {
+    // If ids is a Set (i.e. returned from Queue._scanForJobs), convert it to an Array
+    const idArray = Array.from(ids);
+    return Promise.all(idArray.map((id) => this.getJob(id)));
+  });
+
+  if (cb) helpers.asCallback(promise, cb);
+  return promise;
 };
 
 Queue.prototype.createJob = function (data) {
   return new Job(this, null, data);
 };
 
 Queue.prototype.getJob = function (jobId, cb) {
-  var self = this;
-  if (jobId in this.jobs) {
-    return process.nextTick(cb.bind(null, null, this.jobs[jobId]));
-  } else {
-    Job.fromId(this, jobId, function (err, job) {
-      /* istanbul ignore if */
-      if (err) return cb(err);
-      self.jobs[jobId] = job;
-      return cb(err, job);
-    });
-  }
+  const promise = this._commandable().then(() => {
+    if (this.jobs.has(jobId)) {
+      return this.jobs.get(jobId);
+    }
+    return Job.fromId(this, jobId);
+  }).then((job) => {
+    if (job && this.settings.storeJobs) {
+      this.jobs.set(jobId, job);
+    }
+    return job;
+  });
+
+  if (cb) helpers.asCallback(promise, cb);
+  return promise;
 };
 
-Queue.prototype.getNextJob = function (cb) {
-  var self = this;
-  this.bclient.brpoplpush(this.toKey('waiting'), this.toKey('active'), 0, function (err, jobId) {
-    /* istanbul ignore if */
-    if (err) return cb(err);
-    return Job.fromId(self, Number(jobId), cb);
+Queue.prototype.removeJob = function (jobId, cb) {
+  const promise = this._evalScript('removeJob', 7,
+    this.toKey('succeeded'), this.toKey('failed'), this.toKey('waiting'), this.toKey('active'),
+    this.toKey('stalling'), this.toKey('jobs'), this.toKey('delayed'), jobId)
+    .then(() => this);
+
+  if (cb) helpers.asCallback(promise, cb);
+  return promise;
+};
+
+Queue.prototype._waitForJob = function () {
+  const idPromise = helpers.deferred(), start = Date.now();
+  let getStart;
+  /* istanbul ignore next */
+  this._verbose('blockingPop');
```

Comment:
do we still need all these? Seems excessive and was only used once for debugging redis